privacy/bolton/loss.py

```diff
@@ -0,0 +1,588 @@
+
+  def lipchitz_constant(self, class_weight):
+    """Lipchitz constant, L
+
+    Args:
+      class_weight: class weights used
+
+    Returns: L
+
+    """
+    raise NotImplementedError("lipchitz constant not implemented for "
+                              "StrongConvex Loss"
+                              "function: %s" % str(self.__class__.__name__))
+
+  def kernel_regularizer(self):
+    """returns the kernel_regularizer to be used. Any subclass should override
+      this method if they want a kernel_regularizer (if required for
+      the loss function to be StronglyConvex
+
+    :return: None or kernel_regularizer layer
+    """
+    return None
+
+  def max_class_weight(self, class_weight, dtype):
+    """the maximum weighting in class weights (max value) as a scalar tensor
+
+    Args:
+      class_weight: class weights used
+      dtype: the data type for tensor conversions.
+
+    Returns: maximum class weighting as tensor scalar
+
+    """
+    class_weight = _ops.convert_to_tensor_v2(class_weight, dtype)
+    return tf.math.reduce_max(class_weight)
+
+
+class StrongConvexHuber(losses.Huber, StrongConvexMixin):
+  """Strong Convex version of Huber loss using l2 weight regularization.
+  """
+
+  def __init__(self,
+               reg_lambda: float,
+               C: float,
+               radius_constant: float,
+               delta: float,
+               reduction: str = losses_utils.ReductionV2.SUM_OVER_BATCH_SIZE,
+               name: str = 'huber',
+               dtype=tf.float32):
+    """Constructor.
+
+    Args:
+      reg_lambda: Weight regularization constant
+      C: Penalty parameter C of the loss term
+      radius_constant: constant defining the length of the radius
+      delta: delta value in huber loss.  When to switch from quadratic to
+            absolute deviation.
+      reduction: reduction type to use. See super class
+      name: Name of the loss instance
+      dtype: tf datatype to use for tensor conversions.
+
+    Returns:
+      Loss values per sample.
+    """
+    if C <= 0:
+      raise ValueError('c: {0}, should be >= 0'.format(C))
+    if reg_lambda <= 0:
+      raise ValueError("reg lambda: {0} must be positive".format(reg_lambda))
+    if radius_constant <= 0:
+      raise ValueError('radius_constant: {0}, should be >= 0'.format(
+          radius_constant
+      ))
+    self.C = C
+    self.radius_constant = radius_constant
+    self.dtype = dtype
+    self.reg_lambda = tf.constant(reg_lambda, dtype=self.dtype)
+    super(StrongConvexHuber, self).__init__(
+        delta=delta,
+        name=name,
+        reduction=reduction,
+    )
+
+  def call(self, y_true, y_pred):
+    """Compute loss
+
+    Args:
+      y_true: Ground truth values.
+      y_pred: The predicted values.
+
+    Returns:
+      Loss values per sample.
+    """
+    # return super(StrongConvexHuber, self).call(y_true, y_pred) * self._sample_weight
+    h = self._fn_kwargs['delta']
+    z = y_pred * y_true
```

Comment:
Assuming y_true is label encoded, the formula requires one-hot encode y_true with (-1 for negative and 1 for positive)